WordPress/wordpress-playground
# Draft: Support dynamic loading of PHP extensions via dlopen

## What is this PR doing?

Explores support for dynamic loading of PHP extensions.

Related to #314, #655, #89

[Minimal dlopen example without PHP involved.](https://gist.github.com/adamziel/9c7d3302407294e12d9c1d1eaa54cdba)

## Testing instructions

Clone this branch to a directory with path `/Users/cloudnik/www/Automattic/core/plugins/playground-2/wordpress-playground`. It's not great, I know. I want to remove the explicit path dependency soon.

Then, create a `/usr/local/etc/php.ini` file with the following contents:

```ini
allow_url_fopen=1
extension=/Users/cloudnik/www/Automattic/core/plugins/playground-2/wordpress-playground/packages/php-wasm/node/public/mbstring.so
zend_extension=/Users/cloudnik/www/Automattic/core/plugins/playground-2/wordpress-playground/.libs/xdebug.so
xdebug.mode=debug
sys_temp_dir=/tmp

```

Finally, run:

```
$ nx reset; PHP=8.2 nx start php-wasm-cli -z ./.libs/xdebug.so ./test.php
```

Here's what you should see:

<img width="579" alt="CleanShot 2023-12-01 at 12 44 58@2x" src="https://github.com/WordPress/wordpress-playground/assets/205419/fec08092-f8b2-41b2-b304-9cf8fbdc7a1f">


## Remaining work

- [ ] Make this work with any disk path, not just the one listed above
- [ ] Clean up the code
- [ ] Make the code modular so that PHP extensions may be loaded via a configuration option
- [ ] Patch the final `php.js` files as mentioned in the comments below
- [ ] Ship the following extensions and their required libraries as `.so` files:
   - [x] mbstring
   - [x] gd
   - [ ] iconv
   - [ ] xdebug (optionally)
- [ ] Remove unused exports from the final `php.js` file (they inflate the code from 200KB to 5MB)
- [ ] Provide a way of requesting specific PHP extension on the web and in Node.js

## Follow-up work

* Remove unused symbols from shared libraries (like libpng)
* Explore a Make-based build system to avoid repeated rebuilding of the same libraries
